### PR TITLE
Bugfix/capsys with cli logging

### PR DESCRIPTION
**Pull request by Sup3rGeo**
Fixes https://github.com/pytest-dev/pytest/issues/3819

- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](/changelog/README.rst) for details.
- [x] Target the `master` branch for bug fixes, documentation updates and trivial changes.
- [x] Target the `features` branch for new features and removals/deprecations.
- [x] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

Unless your change is trivial or a small documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Add yourself to `AUTHORS` in alphabetical order;

### CODE REVIEW

File: src/_pytest/capture.py [hunk 1]
```diff
@@ -121,6 +122,25 @@ def suspend_global_capture(self, item=None, in_=False):
                 cap.suspend_capturing(in_=in_)
             return outerr
 
+    @contextlib.contextmanager
+    def disabled(self):
+        """Temporarily disables capture while inside the 'with' block."""
+        if self._current_item is None:
```
Comment (asottile) [inline src/_pytest/capture.py]: can probably combine with the below and do:

```python
if getattr(self._current_item, '_capture_fixture', None) is None:
    yield
else:
    # use self._current_item._capture_fixture directly 
    ...
```

File: src/_pytest/logging.py [hunk 2]
```diff
@@ -573,8 +573,11 @@ def set_when(self, when):
 
     def emit(self, record):
         if self.capture_manager is not None:
-            self.capture_manager.suspend_global_capture()
-        try:
+            ctx_manager = self.capture_manager.disabled()
+        else:
+            ctx_manager = _dummy_context_manager()
+
+        with ctx_manager:
```
Comment (asottile) [inline src/_pytest/logging.py]: nice, I love me some context managers!

**Comment by nicoddemus** [comment]
Hi @Sup3rGeo, hold on please, I'm working on this branch with a few minor improvements which I decided to just go ahead with, if you are OK with that. 👍

**Comment by nicoddemus** [comment]
Made some tweaks, let me know what you think:

* Set `self._current_item = None` at the end of the `pytest_runtest_*` calls to avoid leaking the last item in the run.
* Moved the `dummy_context_manager` to `_pytest.compat` so we can reuse it in `capture` and `logging`.
* Renamed `suspend` to `_suspend` in `CaotureFixture` as that's an internal API.
* Renamed `CaptureManager.disabled` to `CaptureManager.global_and_fixture_disabled` to better convey its meaning.

**Comment by Sup3rGeo** [comment]
That's great @nicoddemus, thanks!

Just one little question about the capture - why do we have both a global and a fixture capture?

This is related to your last item - will we ever want to disable global capture without disabling also the fixture captures? As at least, as I understand, the fixture capture is also part of the whole capture plugin (and CaptureManager.disable would just mean to disable the capture plugin altogether)?

But this is really not very important, I agree with merging the way it is now :)

**Comment by Sup3rGeo** [comment]
> I think it is OK to break the abstraction and be very explicit at what it does, so the caller is well aware that we are disabling global and fixture capturing with it

Thanks for the explanation! It makes sense.

So if I understood correctly the purpose of the capsys/etc fixtures is really to steal the capture even from pytest itself. Otherwise we could just have provided tests access to the global capture from a fixture I guess.

**Comment by nicoddemus** [comment]
> So if I understood correctly the purpose of the capsys/etc fixtures is really to steal the capture even from pytest itself. Otherwise we could just have provided tests access to the global capture from a fixture I guess.

One of the reasons is that you can change capture methods: execute with `--capture=sys` and still use the `capfd` fixture.

**Comment by nicoddemus** [comment]
Awesome, thanks @Sup3rGeo!

**Pull request merged**